src/dune_rules/odoc_new.ml

```diff
@@ -0,0 +1,1959 @@
+              | Ok l -> l
+              | Error _ ->
+                Log.info
+                  [ Pp.textf "Error adding external dir %s to categorized map" top_dir ];
+                cats.external_dirs
+            in
+            Memo.return { cats with external_dirs })
+    in
+    let module Input = struct
+      type t = Context.t * bool
+
+      let equal (c1, b1) (c2, b2) = Context.equal c1 c2 && b1 = b2
+      let hash (c, b) = Poly.hash (Context.hash c, b)
+      let to_dyn _ = Dyn.Opaque
+    end
+    in
+    Memo.create "categorized" ~input:(module Input) run
+  ;;
+
+  let get_categorized ctx all = Memo.exec get_categorized_memo (ctx, all)
+end
+
+module Dep : sig
+  (** [html_alias ctx target] returns the alias that depends on all html targets
+      produced by odoc for [target] *)
+  val html_alias : Path.Build.t -> Alias.t
+
+  (** [deps ctx all findlib_paths valid_libs pkg requires] returns all odoc
+      dependencies of [requires]. If a package [pkg] is also specified, then
+      the odoc dependencies of the package are returned - these are the odoc
+      files representing mld files in the package. [findlib_paths] is the
+      result of [Context.findlib_paths] and [valid_libs] comes from the
+      [Valid] module above. *)
+  val deps
+    :  Context.t
+    -> all:bool
+    -> Path.t list
+    -> Lib.t list
+    -> string option
+    -> Lib.t list Resolve.t
+    -> unit Action_builder.t
+
+  (*** [setup_deps ctx all index odocs] Adds [odocs] as dependencies for [index].
+    These dependencies may be used using the [deps] function *)
+  val setup_deps : Context.t -> all:bool -> Index.t -> Path.Set.t -> unit Memo.t
+end = struct
+  let html_alias dir = Alias.make Alias0.doc_new ~dir
+  let alias = Alias.make (Alias.Name.of_string ".odoc-all")
+
+  let deps ctx ~all findlib_paths valid_libs pkg requires =
+    let open Action_builder.O in
+    let* libs = Resolve.read requires in
+    Action_builder.deps
+      (let init =
+         match pkg with
+         | None -> Dep.Set.empty
+         | Some p ->
+           let index = [ Index.Package (Unknown, p) ] in
+           Dep.Set.singleton (Dep.alias (alias ~dir:(Index.odoc_dir ctx ~all index)))
+       in
+       List.fold_left libs ~init ~f:(fun acc (lib : Lib.t) ->
+         match List.mem ~equal:lib_equal valid_libs lib with
+         | false -> acc
+         | true ->
+           let index =
+             match Lib.Local.of_lib lib with
+             | None -> Index.of_external_lib findlib_paths Unknown lib
+             | Some _ -> Index.of_local_lib lib
+           in
+           let dir = Index.odoc_dir ctx ~all index in
+           let alias = alias ~dir in
+           Dep.Set.add acc (Dep.alias alias)))
+  ;;
+
+  let alias ctx ~all index = alias ~dir:(Index.odoc_dir ctx ~all index)
+
+  let setup_deps ctx ~all m files =
+    Rules.Produce.Alias.add_deps (alias ctx ~all m) (Action_builder.path_set files)
+  ;;
+end
+
+let contains_double_underscore s =
```

Comment:
Which packages require this hack to work?

Reply:
Dream is one - the sub libraries are named with double underscores.

Reply:
Oh, and eio too: ocaml/odoc#1013